File: csharp/Betfair.ESAClient/Betfair.ESAClient/Cache/Market.cs
using Betfair.ESASwagger.Model;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Cache
{
    /// <summary>
    /// Thread safe, reference invariant reference to a market.
    /// Repeatedly calling <see cref="Snap"/> will return atomic snapshots of the market.
    /// </summary>
    public class Market
    {
        private readonly MarketCache _marketCache;
        private readonly string _marketId;
        private readonly Dictionary<RunnerId, MarketRunner> _marketRunners = new Dictionary<RunnerId, MarketRunner>();
        private MarketDefinition _marketDefinition;
        private double _tv;
        private MarketSnap _snap;

        public Market(MarketCache marketCache, string marketId)
        {
            _marketCache = marketCache;
            _marketId = marketId;
        }

        internal void OnMarketChange(MarketChange marketChange)
        {
            //initial image means we need to wipe our data
            bool isImage = marketChange.Img == true;

            if (marketChange.MarketDefinition != null)
            {
                //market definition changed
                OnMarketDefinitionChange(marketChange.MarketDefinition);
            }
            if (marketChange.Rc != null)
            {
                //runners changed
                foreach (RunnerChange runnerChange in marketChange.Rc)
                {
                    OnPriceChange(isImage, runnerChange);
                }
            }

            MarketSnap newSnap = new MarketSnap();
            newSnap.MarketId = _marketId;
            newSnap.MarketDefinition = _marketDefinition;
            newSnap.MarketRunners = _marketRunners.Values.Select(runner => runner.Snap).ToList();
            newSnap.TradedVolume = Utils.SelectPrice(isImage, ref _tv, marketChange.Tv);
            _snap = newSnap;
        }

        private MarketRunner GetOrAdd(RunnerId rid)
        {
            MarketRunner runner;
            if (!_marketRunners.TryGetValue(rid, out runner))
            {
                runner = new MarketRunner(this, rid);
                _marketRunners[rid] = runner;
            }
            return runner;
        }

        private void OnPriceChange(bool isImage, RunnerChange runnerChange)
        {
            MarketRunner marketRunner = GetOrAdd(new RunnerId(runnerChange.Id, runnerChange.Hc));
            //update the runner
            marketRunner.OnPriceChange(isImage, runnerChange);
        }

        private void OnMarketDefinitionChange(MarketDefinition marketDefinition)
        {
            _marketDefinition = marketDefinition;
            if (marketDefinition.Runners != null)
            {
                foreach (RunnerDefinition runnerDefinition in marketDefinition.Runners)
                {
                    OnRunnerDefinitionChange(runnerDefinition);
                }
            }
        }

        private void OnRunnerDefinitionChange(RunnerDefinition runnerDefinition)
        {
            MarketRunner marketRunner = GetOrAdd(new RunnerId(runnerDefinition.Id, runnerDefinition.Hc));
            //update runner
            marketRunner.OnRunnerDefinitionChange(runnerDefinition);
        }

        /// <summary>
        /// Market id
        /// </summary>
        public string MarketId
        {
            get
            {
                return _marketId;
            }
        }

        /// <summary>
        /// Whether the market is closed.
        /// </summary>
        public bool IsClosed
        {
            get
            {
                return _marketDefinition != null && _marketDefinition.Status == MarketDefinition.StatusEnum.Closed;
            }
        }

        /// <summary>
        /// An atomic snapshot of the state of the market.
        /// </summary>
        public MarketSnap Snap
        {
            get
            {
                return _snap;
            }
        }

        public override string ToString()
        {
            return "Market{" +
                "MarketId=" + MarketId +
                ", MarketDefinition=" + _marketDefinition +
                ", MarketRunners=" + String.Join(", ", _marketRunners.Values) +
                "}";
        }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/cache/market/Market.java
package com.betfair.esa.client.cache.market;

import com.betfair.esa.client.cache.util.RunnerId;
import com.betfair.esa.client.cache.util.Utils;
import com.betfair.esa.swagger.model.MarketChange;
import com.betfair.esa.swagger.model.MarketDefinition;
import com.betfair.esa.swagger.model.RunnerChange;
import com.betfair.esa.swagger.model.RunnerDefinition;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Thread safe, reference invariant reference to a market. Repeatedly calling <see cref="Snap"/>
 * will return atomic snapshots of the market.
 */
public class Market {
    private final MarketCache marketCache;
    private final String marketId;
    private final Map<RunnerId, MarketRunner> marketRunners = new ConcurrentHashMap<>();
    private MarketDefinition marketDefinition;
    private double tv;
    // An atomic snapshot of the state of the market.
    private MarketSnap snap;

    public Market(MarketCache clientMarketCache, String marketId) {
        this.marketCache = clientMarketCache;
        this.marketId = marketId;
    }

    void onMarketChange(MarketChange marketChange) {
        // initial image means we need to wipe our data
        boolean isImage = Boolean.TRUE.equals(marketChange.isImg());
        // market definition changed
        Optional.ofNullable(marketChange.getMarketDefinition())
                .ifPresent(this::onMarketDefinitionChange);
        // runners changed
        Optional.ofNullable(marketChange.getRc())
                .ifPresent(l -> l.forEach(p -> onPriceChange(isImage, p)));

        MarketSnap newSnap = new MarketSnap();
        newSnap.setMarketId(marketId);
        newSnap.setMarketDefinition(marketDefinition);
        newSnap.setMarketRunners(
                marketRunners.values().stream()
                        .map(MarketRunner::getSnap)
                        .collect(Collectors.toList()));
        newSnap.setTradedVolume(tv = Utils.selectPrice(isImage, tv, marketChange.getTv()));
        snap = newSnap;
    }

    private void onPriceChange(boolean isImage, RunnerChange runnerChange) {
        MarketRunner marketRunner =
                getOrAdd(new RunnerId(runnerChange.getId(), runnerChange.getHc()));
        // update runner
        marketRunner.onPriceChange(isImage, runnerChange);
    }

    private void onMarketDefinitionChange(MarketDefinition marketDefinition) {
        this.marketDefinition = marketDefinition;
        Optional.ofNullable(marketDefinition.getRunners())
                .ifPresent(rds -> rds.forEach(this::onRunnerDefinitionChange));
    }

    private void onRunnerDefinitionChange(RunnerDefinition runnerDefinition) {
        MarketRunner marketRunner =
                getOrAdd(new RunnerId(runnerDefinition.getId(), runnerDefinition.getHc()));
        // update runner
        marketRunner.onRunnerDefinitionChange(runnerDefinition);
    }

    private MarketRunner getOrAdd(RunnerId runnerId) {
        MarketRunner runner =
                marketRunners.computeIfAbsent(runnerId, k -> new MarketRunner(this, k));
        return runner;
    }

    public String getMarketId() {
        return marketId;
    }

    public boolean isClosed() {
        // whether the market is closed
        return (marketDefinition != null
                && marketDefinition.getStatus() == MarketDefinition.StatusEnum.CLOSED);
    }

    public MarketSnap getSnap() {
        return snap;
    }

    @Override
    public String toString() {
        return "Market{"
                + "marketId='"
                + marketId
                + '\''
                + ", marketRunners="
                + marketRunners
                + ", marketDefinition="
                + marketDefinition
                + '}';
    }
}
